spec/assessments/KeywordDensityAssessmentSpec.js

```diff
@@ -1,105 +1,168 @@
+		expect( result.getText() ).toBe( "<a href='https://yoa.st/33v' target='_blank'>Keyphrase density</a>: 2%. This is great!" );
+	} );
 
-		paper = new Paper( "string with the keyword", { keyword: "keyword" } );
-		result = new KeywordDensityAssessment().getResult( paper, factory.buildMockResearcher( {
-			getKeywordDensity: 10,
-			keywordCount: {
-				count: 1,
-			},
-		}, true ), i18n );
+	it( "runs the keywordDensity on the paper with a slightly too high keyphrase density (3%)", function() {
+		const paper = new Paper( mediumText + mediumText + "a keyword. Keyword. Keyword. ", { keyword: "keyword" } );
+		const researcher = new Researcher( paper );
+		const result = new KeywordDensityAssessment().getResult( paper, researcher, i18n );
+		expect( result.getScore() ).toBe( -10 );
+		expect( result.getText() ).toBe( "<a href='https://yoa.st/33v' target='_blank'>Keyphrase density</a>: 3%. This is over the advised 2.5% maximum; the keyphrase was found 3 times. <a href='https://yoa.st/33w' target='_blank'>Don't overoptimize</a>!" );
+	} );
+
+	it( "runs the keywordDensity on the paper with a very high keyphrase density (10%)", function() {
+		const paper = new Paper( shortText + "keyword", { keyword: "keyword" } );
+		const researcher = new Researcher( paper );
+		const result = new KeywordDensityAssessment().getResult( paper, researcher, i18n );
 		expect( result.getScore() ).toBe( -50 );
 		expect( result.getText() ).toBe( "<a href='https://yoa.st/33v' target='_blank'>Keyphrase density</a>: 10%. This is way over the advised 2.5% maximum; the keyphrase was found 1 time. <a href='https://yoa.st/33w' target='_blank'>Don't overoptimize</a>!" );
+	} );
+} );
 
-		paper = new Paper( "string with the keyword", { keyword: "keyword" } );
-		result = new KeywordDensityAssessment().getResult( paper, factory.buildMockResearcher( {
-			getKeywordDensity: 2,
-			keywordCount: {
-				count: 1,
-			},
-		}, true ), i18n );
+describe( "Tests for the keywordDensity assessment for languages with morphology, for regular analysis", function() {
+	beforeEach( () => {
+		process.env.YOAST_RECALIBRATION = "disabled";
+	} );
+	it( "gives a GOOD result when keyword density is 3.0%", function() {
+		const paper = new Paper( mediumText + mediumText + "a keyword. Keyword. Keyword. ", { keyword: "keyword", locale: "en_EN" } );
+		const researcher = new Researcher( paper );
+		researcher.addResearchData( "morphology", morphologyData );
+		const result = new KeywordDensityAssessment().getResult( paper, researcher, i18n );
 		expect( result.getScore() ).toBe( 9 );
-		expect( result.getText() ).toBe( "<a href='https://yoa.st/33v' target='_blank'>Keyphrase density</a>: 2%. This is great!" );
+		expect( result.getText() ).toBe( "<a href='https://yoa.st/33v' target='_blank'>Keyphrase density</a>: 3%. This is great!" );
+	} );
+} );
+
+describe( "Tests for the keywordDensity assessment for languages without morphology, for recalibration analysis", function() {
```

Comment:
Might be a good idea to add some tests with multi-word keyphrases (while keeping the overall word count the same) to see that the assessment returns different results when the keyphrase length varies.

Reply:
Fixed.